x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,1301 @@
+-   [List of things to fix (create tickets)](#list-of-things-to-fix-create-tickets){#toc-list-of-things-to-fix-create-tickets}
+    -   [Rule fields](#rule-fields){#toc-rule-fields}
+-   [Updating Prebuilt Rules](#updating-prebuilt-rules){#toc-updating-prebuilt-rules}
+    -   [Changes to upgrade `_review` and `_perform` endpoints](#changes-to-upgrade-_review-and-_perform-endpoints){#toc-changes-to-upgrade-_review-and-_perform-endpoints}
+    -   [Rule field diff algorithms](#rule-field-diff-algorithms){#toc-rule-field-diff-algorithms}
+-   [Changes in UI](#changes-in-ui){#toc-changes-in-ui}
+-   [Scenarios for bulk accepting updates](#scenarios-for-bulk-accepting-updates){#toc-scenarios-for-bulk-accepting-updates}
+-   [Marking rules (and their fields) as customized](#marking-rules-and-their-fields-as-customized){#toc-marking-rules-and-their-fields-as-customized}
+-   [Other open questions](#other-open-questions){#toc-other-open-questions}
+
+
+
+## Necessary rule schema changes
+
+In order to support the customization of Elastic Prebuilt Rules, we need to modify our rule schema. This involves introducing two new fields: `prebuilt` and `customized`. Simultaneously, we need to deprecate the existing `immutable` field, which determines if a rule is an Elastic Prebuilt Rule and whether its fields can be modified/customized.
+
+To ensure backward compatibility and avoid breaking changes, we will silently deprecate the `immutable` field but keep it unchanged (i.e. non-optional).
+
+
+In summary, the changes to be done to our rule schema are:
+
+- Mark the `immutable` field as deprecated but keep unchanged for backwards compatibility.
+    - Should remain a `boolean` type and non-optional (`required`)
+- Add `prebuilt` field to the rule schema
+    - `prebuilt` should have same type as `immutable`: `boolean`
+    - Should be non-optional (`required`)
+    - During migration, set the `prebuilt` value to match the current value of `immutable`. This means: prebuilt rules should be migrated with a `prebuilt` value set to `true`, while custom rules should be set with a `prebuilt` value of `false`.
+- Add `customized` field to the rule schema
+    - Should have type of `boolean`
+    - Should be non-optional (`required`)
+    - If non existing, initialize it with a default value of `false` for all rules, both custom and prebuilt. The only exception for this is when importing prebuilt rules (see below).
+    - Custom rules (`prebuilt: false`) will never use this field, and its value should always be `false`.
+    - When prebuilt rules are installed for the first time, set the `customized` field to `false`. Any update to the rule's cutomizable fields that causes it to diverge from the base version should, as a side effect, update the `customized` field to `true`.
+
+
+```ts
+// rule_schema/model/rule_schema.ts
+
+[...]
+export type BaseRuleParams = z.infer<typeof BaseRuleParams>;
+export const BaseRuleParams = z.object({
```

Comment:
If we're doing migration on write, we need to differentiate between changes to the internal rule schema and the API schema. Migration on write means that the internal rule schema needs to represent that the new fields may not always exist, so `prebuilt` and `customized` must be optional. In the API schema, those fields can be required if we're comfortable with the breaking change. The conversion layer between internal and API schema will be responsible for taking rule objects where those fields may not exist and populating the correct value in the API response where they must exist.

This schema here appears to be the internal rule schema, which should not have `prebuilt` and `customized` listed as required if we're doing migration on write.